pkg/tipselect/urts.go

```diff
@@ -0,0 +1,485 @@
+	// Events are the events that are triggered by the TipSelector.
+	Events Events
+}
+
+// New creates a new tip-selector.
+func New(maxDeltaTxYoungestRootSnapshotIndexToLSMI int, maxDeltaTxApproveesOldestRootSnapshotIndexToLSMI int, belowMaxDepth int, maxReferencedTipAgeSeconds time.Duration, maxApprovers uint32) *TipSelector {
+	return &TipSelector{
+		maxDeltaTxYoungestRootSnapshotIndexToLSMI:        milestone.Index(maxDeltaTxYoungestRootSnapshotIndexToLSMI),
+		maxDeltaTxApproveesOldestRootSnapshotIndexToLSMI: milestone.Index(maxDeltaTxApproveesOldestRootSnapshotIndexToLSMI),
+		belowMaxDepth:              milestone.Index(belowMaxDepth),
+		maxReferencedTipAgeSeconds: maxReferencedTipAgeSeconds,
+		maxApprovers:               maxApprovers,
+		tipsMap:                    make(map[string]*Tip),
+		Events: Events{
+			TipAdded:        events.NewEvent(TipCaller),
+			TipRemoved:      events.NewEvent(TipCaller),
+			TipSelPerformed: events.NewEvent(WalkerStatsCaller),
+		},
+	}
+}
+
+// AddTip adds the given tailTxHash as a tip.
+func (ts *TipSelector) AddTip(tailTxHash hornet.Hash) {
+	ts.tipsLock.Lock()
+	defer ts.tipsLock.Unlock()
+
+	if _, exists := ts.tipsMap[string(tailTxHash)]; exists {
+		// tip already exists
+		return
+	}
+
+	lsmi := tangle.GetSolidMilestoneIndex()
+
+	score := ts.calculateScore(tailTxHash, true, lsmi)
+	if score == ScoreLazy {
+		// do not add lazy tips.
+		// lazy tips should also not remove other tips from the pool.
+		return
+	}
+
+	tip := &Tip{
+		Score:             score,
+		Hash:              tailTxHash,
+		TimeFirstApprover: time.Time{},
+		ApproversCount:    atomic.NewUint32(0),
+	}
+	ts.tipsMap[string(tailTxHash)] = tip
+	ts.scoreSum += int(tip.Score)
+
+	ts.Events.TipAdded.Trigger(tip)
+
+	// search all referenced tails of this Tip and remove them from the tip pool
+	approveeTailTxHashes, err := dag.FindAllTails(tailTxHash, true)
+	if err != nil {
+		return
+	}
+
+	for approveeTailTxHash := range approveeTailTxHashes {
+		if approveeTip, exists := ts.tipsMap[approveeTailTxHash]; exists {
+			// check if the maximum amount of approvers for this tip is reached
+			if approveeTip.ApproversCount.Add(1) >= ts.maxApprovers {
+				ts.removeTipWithoutLocking(hornet.Hash(approveeTailTxHash))
+				continue
+			}
+
+			// check if the tip was referenced by another transaction before
+			if approveeTip.TimeFirstApprover.IsZero() {
+				approveeTip.TimeFirstApprover = time.Now()
+
+				// remove the tip after it reaches its maximum age
+				time.AfterFunc(ts.maxReferencedTipAgeSeconds, func() {
+					ts.RemoveTip(hornet.Hash(approveeTailTxHash))
+				})
+			}
+		}
+	}
+}
+
+// removeTipWithoutLocking removes the given tailTxHash from the tipsMap without acquiring the lock.
+func (ts *TipSelector) removeTipWithoutLocking(tailTxHash hornet.Hash) {
+	if tip, exists := ts.tipsMap[string(tailTxHash)]; exists {
+		ts.scoreSum -= int(tip.Score)
+		delete(ts.tipsMap, string(tailTxHash))
+		ts.Events.TipRemoved.Trigger(tip)
```

Comment:
sure we should trigger an event here while locked?

Reply:
Do you have an idea how to do that, without making everything unreadable and complex? I unfortunately don't have one :(